### DISCUSSION TITLE
Installation without conda and conan

### DISCUSSION THREAD

**Discussion by LovisAnderson** [ds4dm/ecole#121] [Questions] [answered]
Hi,

thanks for developing and sharing ecole! I am trying to install ecole on a HPC environment, where conda is sadly not an option. When trying to install it from source, the package _conan_ is problematic, as it has dependencies, that themselves have dependencies to old software that won't compile.

Is it possible to build ecole without conan using CMake functionalities?

Any help would be very much appreciated. Have a nice day!

Lovis

**Comment by AntoinePrv** [comment, marked as answer]
Hi Lovis,

We are aware that some people (including ourselves) need to use Ecole without Conda. Proper support is in our todo list.

Ecole has some compiled C++, so when we talk about Ecole dependencies, we need to distinguish two types:
  - *Build dependencies* are needed to compile Ecole. For us, it's anything from compilers, CMake, Conan,... to libraries.
  - *Runtime dependencies* need to be present to use Ecole. Currently the only library that is needed when running Ecole are SCIP and Python.
  
 ## Conda
 Conda is great for us, because it tackles both aspects. It ships Ecole already compiled, and resolves SCIP as a dependency.
 
 ## Source installation
 To compile, you would need to get a recent compiler, CMake, Make, Xtensor/Xtl/Xtensor-Python, PyBind, Python, Numpy, everything listed in `conanfile.txt`, and SCIP.
 Regarding Conan, we can indeed make it optional, and use CMake to get the source libraries.
 
Alternatively, you do not need to compile Ecole on the same machine that you will be running it.
This is what happens with conda, it is compiled once and distributed to many.
However, if you decide to go that way, you need to make SCIP available and findable by Ecole (which we have not dug into so far).

## Pip
This last idea of compiling Ecole and shipping without SCIP is the first step we want to explore with Pip.
`pip install ecole` get Ecole already compiled, but users would still need to point it at the correct SCIP (or install SCIP in a default location).

## What's the solution?
In our HPC servers (ComputeCanada), Conda is also not recommended (this is not entirely justified IMHO).
However, we decided that Conda was still the best option for us.

If Conan is the only thing getting in your way, we can replace it with CMake's `FetchContent` utilities.
You would still need to provide you own SCIP, and get a C++17 compiler.

Let us know what you think and what approaches could work for you.

**Comment by AntoinePrv** [comment]
Out of curiosity, what errors are you getting with conan?

**Reply by LovisAnderson** [reply]
Hallo Antoine,

with your help we managed to install ecole on our HPC environment! Many thanks! We had problems with the conan (version1.33.0) dependency to fas-0.1.5 which requires PyYAML==3.10. The compilation of this version fails with the following issue (solved in version 3.13): https://github.com/yaml/pyyaml/issues/126
 
In addition, we changed the following dependencies of fas so that the required dependencies are compatible with other conan-dependencies:
```
-        'PyYAML==3.10',
+        'PyYAML==3.13',
-        'requests==2.7.0',
+        'requests>=2.7.0',
-        'six==1.10.0',
+        'six==1.15.0',
```

With these dependencies everything seems to work so far.

Best
   Lovis

**Comment by AntoinePrv** [comment]
As of version `v0.5.0`, Ecole is now much easier to install without Conda and Conan. See the [`v0.5.x` instructions](https://doc.ecole.ai/v0.5.x/installation.html#from-source).

**Reply by thobreuer** [reply]
Thanks for adapting the installation procedure. It works like a charm :)
